operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java

```diff
@@ -93,9 +103,36 @@ public Map<String, EventSource> prepareEventSources(EventSourceContext<Keycloak>
         sources.put("serviceSource", servicesEvent);
         sources.putAll(EventSourceInitializer.nameEventSources(statefulSetEvent,
                 watchedSecrets.getWatchedSecretsEventSource()));
+
+        if (context.getClient().supports(Ingress.class)) {
+            watchOpenShiftIngress(context, sources);
+        }
+
         return sources;
     }
 
+    private void watchOpenShiftIngress(EventSourceContext<Keycloak> context, Map<String, EventSource> sources) {
+        InformerConfiguration<Ingress> ingressIC = InformerConfiguration.from(Ingress.class)
+                .withOnUpdateFilter((oldResource,
+                        newResource) -> CLUSTER_CONFIG_NAME.equals(newResource.getMetadata().getName())
+                                && (!Objects.equals(newResource.getSpec().getAppsDomain(), oldResource.getSpec().getAppsDomain()))
+                                || !Objects.equals(newResource.getSpec().getDomain(), oldResource.getSpec().getDomain()))
+                .withPrimaryToSecondaryMapper(primary -> Set.of(new ResourceID(CLUSTER_CONFIG_NAME)))
+                .withSecondaryToPrimaryMapper(dependentResource -> {
+                    if (!CLUSTER_CONFIG_NAME.equals(dependentResource.getMetadata().getName())) {
+                        return Collections.emptySet();
+                    }
+                    return context.getPrimaryCache().list()
+                            .filter(kc -> (kc.getSpec().getIngressSpec() == null
+                                    || kc.getSpec().getIngressSpec().isIngressEnabled())
+                                    && Optional.ofNullable(kc.getSpec().getHostnameSpec())
+                                            .map(HostnameSpec::getHostname).isEmpty())
+                            .map(ResourceID::fromResource).collect(Collectors.toSet());
+                }).build();
+        EventSource ingressEvent = new InformerEventSource<>(ingressIC, context);
+        sources.putAll(EventSourceInitializer.nameEventSources(ingressEvent));
+    }
```

Comment:
This feels to me like maybe a bit too much of added complexity for only a potential corner case when the root hostname for the cluster changes.

If we just directly set the generated hostname on the CR spec (i.e. treat it as a default value that gets persited with the CR), we don't need this.

Reply:
Not entirely, you still need the informer to avoid repeated lookups of the hostname - unless you want to simply cache the first successful lookup.

> If we just directly set the generated hostname on the CR spec (i.e. treat it as a default value that gets persited with the CR), we don't need this.

I had considered that, but it unfortunately also complicates the logic as we'll still need to make the decision to persist the keycloak up in the KeycloakController.

Reply:
Double checked the openshift logic - there's no expectation that existing routes will change if the appsDomain is changed, but newly created routes will have the updated appsDomain.  The user is expected to update existing ones.  That gets a little messy in our case - if we cache the appsDomain in our operator not via an informer that will require an operator restart to use the new one.  If we're not watching the ingress config, then if we populate the hostname someone will have to remove it, or set it to the new expected value.  If we don't populate the hostname and aren't watching, then something will need to trigger another reconciliation.  

To clarify then do you not want to use an informer at all, or just don't try to react to appDomain changes wrt existing instances?